tools/testing/e2e/iree-e2e-matmul-test.cc

```diff
@@ -205,20 +205,114 @@ static iree_status_t reference_matmul(
-                  m_size, k_size, n_size, lhs_type, rhs_type, acc_type,
-                  transpose_rhs, lhs_contents.data, rhs_contents.data,
-                  acc_contents.data, result_contents.data, m, n));
+  // This function computes:
+  //
+  // C = matmul(A, B) + D
+  //
+  // where the correspond to function arguments is:
+  //        C - result_contents.
+  //        A - lhs_contents.
+  //        B - rhs_contents.
+  //        D - acc_contents.
+  //
+  // Because the rows of A and the columns of B repeat,
+  //
+  // C[m0, n0] == C[m1, n1] if
+  //  - m0 % m_repeat_period == m1 % m_repeat_period, and
+  //  - n0 % n_repeat_period == n1 % n_repeat_period.
+  //
+  // This cyclical repetition of the values in C can be used to accelerate the
+  // computation of C: By computing C[i, j] for all
+  //   i <= m_repeat_period, and
+  //   j <= n_repeat_period,
+  //
+  // all remaining values of C can be copied from these low-index values,
+  // avoiding computing the O(k) inner-product and instead performing an O(1)
+  // copy.
+  //
+  // Using the cyclical caching trick to compute only some of the results when
+  // 'compute_every > 1' is too complicated, unless an additional buffer is used
+  // to recored which elements have previously been computed (which we don't
+  // want to do). We therefore use only one of the two techniques to reduce
+  // computation.
+
+  // Repeat period of 0 denotes no repeat.
+  m_repeat_period = m_repeat_period > 0 ? m_repeat_period : m_size;
+  n_repeat_period = n_repeat_period > 0 ? n_repeat_period : n_size;
+
+  // The number of O(k) inner-products that will be computed using the caching
+  // trick. Only inner-products for
+  // (i,j) in {0, ..., m_repeat_period-1} x {0, ..., n_repeat_period-1}
+  // computed.
+  const int64_t caching_trick_cost = m_repeat_period * n_repeat_period;
+
+  // The number of O(k) inner-products that will be computed by computing only
+  // every compute_every-th element.
+  const int64_t compute_every_cost = m_size * n_size / compute_every;
+
+  // True if the repeat period in both the m- and n- dimensions are too large to
+  // result in any caching.
+  const bool no_cached_values =
+      m_repeat_period >= m_size && n_repeat_period >= n_size;
+
+  // Even if compute_every > 1, we use the caching trick to compute every value
+  // if it is cheaper to do so.
+  if (compute_every > 1 && compute_every_cost < caching_trick_cost) {
+    iree_host_size_t count = 0;
+    for (iree_hal_dim_t m = 0; m < m_size; ++m) {
+      for (iree_hal_dim_t n = 0; n < n_size; ++n) {
+        if (++count < compute_every) continue;
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, reference_matmul_element(
+                    m_size, k_size, n_size, lhs_type, rhs_type, acc_type,
+                    transpose_rhs, lhs_contents.data, rhs_contents.data,
+                    acc_contents.data, result_contents.data, m, n));
+      }
+    }
+  }
+
+  else if (no_cached_values) {
```

Comment:
style:
```suggestion
  } else if (no_cached_values) {
```
(here and elsewhere - keep `} else {` and `} else if(...) {` on the same line)

https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching